### ISSUE TITLE
Unreasonable high GPU usage when using erosion

### ISSUE THREAD

**Issue by rlleshi** [kornia/kornia#2948]
### Describe the bug

Not sure if its a bug. But the GPU usage blows up when using `kornia.morphology.erosion`.

For a tensor of shape `(1, 3, 958, 640)` & kernel of shape `(40, 40)`, the operation is taking ~13G GPU memory on my machine.

### Reproduction steps

```bash
import torch
import kornia as K

def log_gpu_usage(stage: str) -> None:
    print(
        f'{stage} - Allocated: {torch.cuda.memory_allocated() / 1024 ** 2:.2f} MB, '
        f'Cached: {torch.cuda.memory_reserved() / 1024 ** 2:.2f} MB')

device = 'cuda:0'
dummy_tensor = torch.randn((1, 3, 958, 640), device=device)
kernel = torch.ones((40, 40), device=device)
result = K.morphology.erosion(dummy_tensor, kernel)

log_gpu_usage('kornia_test')

Outputs: kornia_test - Allocated: 14.04 MB, Cached: 13218.00 MB
```

### Expected behavior

On average such an operation takes:

- 0.033s on kornia using the GPU
- 2.32s on kornia using the CPU (with RAM memory being significantly occupied)
- 0.046 on opencv using CPU (with RAM memory not seeing any significant occupation)

So kornia on GPU is indeed faster than opencv on CPU. However, the memory usage blows up, and this makes it unsuitable to be used for such cases. 

I understand that further optimization might be tricky and hard. But just wondering if you guys have perhaps already planned it.

### Environment

```shell
❯ python collect_env.py 
Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 23.10 (x86_64)
GCC version: (Ubuntu 12.3.0-9ubuntu2) 12.3.0
Clang version: Could not collect
CMake version: version 3.29.3
Libc version: glibc-2.38

Python version: 3.9.0 (default, Jul 27 2022, 17:09:22)  [GCC 8.4.0] (64-bit runtime)
Python platform: Linux-6.5.0-42-generic-x86_64-with-glibc2.38
Is CUDA available: True
CUDA runtime version: 12.3.52
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2070 Super with Max-Q Design
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 545.29.06
cuDNN version: Probably one of the following:
/usr/lib/cuda/lib64/libcudnn.so.8.9.7
/usr/lib/cuda/lib64/libcudnn_adv_infer.so.8.9.7
/usr/lib/cuda/lib64/libcudnn_adv_train.so.8.9.7
/usr/lib/cuda/lib64/libcudnn_cnn_infer.so.8.9.7
/usr/lib/cuda/lib64/libcudnn_cnn_train.so.8.9.7
/usr/lib/cuda/lib64/libcudnn_ops_infer.so.8.9.7
/usr/lib/cuda/lib64/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
CPU family:                         6
Model:                              165
Thread(s) per core:                 2
Core(s) per socket:                 6
Socket(s):                          1
Stepping:                           2
CPU(s) scaling MHz:                 39%
CPU max MHz:                        5000,0000
CPU min MHz:                        800,0000
BogoMIPS:                           5199,98
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp pku ospke md_clear flush_l1d arch_capabilities
L1d cache:                          192 KiB (6 instances)
L1i cache:                          192 KiB (6 instances)
L2 cache:                           1,5 MiB (6 instances)
L3 cache:                           12 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI Syscall hardening, KVM SW loop
Vulnerability Srbds:                Mitigation; Microcode
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.16.1
[pip3] onnxruntime-gpu==1.18.0
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] Could not collect
```

**Comment by johnnv1** [comment]
hey @rlleshi, thanks for opening an issue for that, it is probably because of the `unfold` operations which are memory hungry, you can use this operator within the argument: `engine='convolution'`, which should be lighter.

From our docs: https://kornia.readthedocs.io/en/stable/morphology.html#kornia.morphology.erosion

In the codebase: https://github.com/kornia/kornia/blob/bdeac07e1edc26863a9c8d0826dc202974fd850a/kornia/morphology/morphology.py#L45

Could you check if it solves your issue?

**Comment by rlleshi** [comment]
@johnnv1 thanks for the quick answer. Unfortunately, that's only ~2G less GPU intensive.

**Comment by Isalia20** [comment]
Please take a look at the PR above. Should drastically reduce the memory usage, 13GB->2GB on your example

**Comment by ducha-aiki** [comment]
Unfortunately, that's how it is. You can try using it with `torch.compile` and usually after  optimization, that is faster on CPU and GPU and uses less memory.

**Comment by minh-nguyenhoang** [comment]
For large binary mask that need erosion/dilation, you can subtitute it with box-blur and thresholding (which box-blur should already be implemented in a really efficient way) - this approach should give you an exact erosion/dilation operator with square/rectangle structuring element. 
Otherwise, you could apply the operator with smaller structuring elements multiple times to achieve some degree of close approximation (if your structuring element is weird enough it can't be created by smaller ones).

**Comment by edgarriba** [comment]
If anyone is interested in this, feel free to re-open the ticket. Thanks!
